patroni/dcs/consul.py

```diff
@@ -324,12 +346,52 @@ def touch_member(self, data, ttl=None, permanent=False):
         try:
             args = {} if permanent else {'acquire': self._session}
             self._client.kv.put(self.member_path, json.dumps(data, separators=(',', ':')), **args)
+            if self._register_service:
+                self.update_service(self._my_member_data, data)
             self._my_member_data = data
             return True
         except Exception:
             logger.exception('touch_member')
         return False
 
+    @catch_consul_errors
+    def register_service(self, service_name, **kwargs):
+        logger.info('Register service {}, params {}'.format(service_name, kwargs))
+        return self.retry(self._client.agent.service.register, service_name, **kwargs)
+
+    def _update_service(self, data):
+        service_name = self._service_name
+        role = data['role']
+        api_parts = urlparse(data['api_url'])
+        api_parts = api_parts._replace(path='/{}'.format(role))
+        conn_parts = urlparse(data['conn_url'])
+        params = {
+            'address': conn_parts.hostname,
+            'port': conn_parts.port,
+            'check': base.Check.http(api_parts.geturl(), self._service_check_interval),
+            'tags': [role]
+        }
+
+        if role in ['master', 'replica']:
```

Comment:
1. I think the best way is to create a special tag for this and make a configuration option to tag this nodes as replica additionally. 
2. Yes, it make sense. Especially with `DeregisterCriticalServiceAfter` it can be deregister by consul and we won't register it again when state change to "running".